ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFNullif.java

```diff
@@ -86,17 +87,13 @@ public ObjectInspector initialize(ObjectInspector[] arguments) throws UDFArgumen
   public Object evaluate(DeferredObject[] arguments) throws HiveException {
     Object arg0 = arguments[0].get();
     Object arg1 = arguments[1].get();
-    Object value0 = null;
-    if (arg0 != null) {
-      value0 = returnOIResolver.convertIfNecessary(arg0, argumentOIs[0], false);
-    }
+    Object value0 = returnOIResolver.convertIfNecessary(arg0, argumentOIs[0], false);
     if (arg0 == null || arg1 == null) {
       return value0;
     }
-    PrimitiveObjectInspector compareOI = (PrimitiveObjectInspector) returnOIResolver.get();
-    if (PrimitiveObjectInspectorUtils.comparePrimitiveObjects(
-        value0, compareOI,
-        returnOIResolver.convertIfNecessary(arg1, argumentOIs[1], false), compareOI)) {
+    Object value1 = returnOIResolver.convertIfNecessary(arg1, argumentOIs[1], false);
+    ObjectInspector compareOI = returnOIResolver.get();
+    if (ObjectInspectorUtils.compare(value0, compareOI, value1, compareOI) == 0) {
```

Comment:
Since now we are expanding the support to non-primitives can/should we support all other categories: MAP, STRUCT, UNION? 
I think we should have positive tests for all supported types and negative tests (exception) for non-supported ones (if any).

Reply:
it should work now; the problem was that `PrimitiveObjectInspectorUtils` was used - but I could add some tests

Reply:
it does work - but there are some interesting findings:

both of the following cases change the "non-used" part of the union (note: `create_union(idx,o0,o1)` creates a union which uses the `idx`-th object)
```
SELECT (NULLIF(create_union(0,1,2),create_union(0,1,3)) is not null);
false
SELECT (NULLIF(create_union(0,1,2),create_union(1,2,1)) is not null);
true
```
it seems like the comparision ignores the unused/dormant parts which could be ok; but it seems like it also compares the `idx` field - I think it should either also ignore the `idx` field or also consider the dormant objects

opened HIVE-25748 to look into this

Reply:
Thanks for adding more tests Zoltan. I always find interesting stuff whenever I decide to add a test :)

FYI: When I worked on adding support for non-primitive types in simple equality operations (HIVE-24886) I excluded UNION types. I wasn't sure what the semantics should be so I though it was better to throw an exception and say the operation is not permitted.

Reply:
I think `UNION` is probably a corner feature and not in wide usage - it's a good idea to throw an exception instead of going down the dragon's path

Reply:
I think the result of union is expected, the `idx` in a union implies [which part of the union is being used](https://cwiki.apache.org/confluence/display/hive/languagemanual+types#LanguageManualTypes-UnionTypesunionUnionTypes) ,  we cannot ignore both the `idx` part in union type.